src/distributions/uniform.rs

```diff
@@ -737,52 +742,75 @@ impl UniformSampler for UniformDuration {
-                UniformDurationMode::Small {
-                    nanos: Uniform::new_inclusive(0, nanos),
-                }
+
+        let low_s = low.as_secs();
+        let low_n = low.subsec_nanos();
+        let mut high_s = high.as_secs();
+        let mut high_n = high.subsec_nanos();
+
+        if high_n < low_n {
+            high_s = high_s - 1;
+            high_n = high_n + 1_000_000_000;
+        } else {
+            high_s = high_s;
+            high_n = high_n;
+        }
+
+        let mode = if low_s == high_s {
+            UniformDurationMode::Small {
+                secs: low_s,
+                nanos: Uniform::new_inclusive(low_n, high_n),
             }
-            None => {
+        } else {
+            let max = high_s
+                .checked_mul(1_000_000_000)
+                .and_then(|n| n.checked_add(high_n as u64));
+
+            if let Some(higher_bound) = max {
+                let lower_bound = low_s * 1_000_000_000 + low_n as u64;
+                UniformDurationMode::Medium {
+                    nanos: Uniform::new_inclusive(lower_bound, higher_bound),
+                }
+            } else {
+                // An offset is applied to simplify generation of nanoseconds
+                let max_nanos = high_n - low_n;
                 UniformDurationMode::Large {
-                    size: size,
-                    secs: Uniform::new_inclusive(0, size.as_secs()),
+                    max_secs: high_s,
+                    max_nanos,
+                    secs: Uniform::new_inclusive(low_s, high_s),
                 }
             }
         };
-
         UniformDuration {
             mode,
-            offset: low,
+            offset: low_n,
         }
     }
 
     #[inline]
     fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Duration {
-        let d = match self.mode {
-            UniformDurationMode::Small { nanos } => {
+        match self.mode {
+            UniformDurationMode::Small { secs, nanos } => {
+                let n = nanos.sample(rng);
+                Duration::new(secs, n)
+            }
+            UniformDurationMode::Medium { nanos } => {
                 let nanos = nanos.sample(rng);
                 Duration::new(nanos / 1_000_000_000, (nanos % 1_000_000_000) as u32)
             }
-            UniformDurationMode::Large { size, secs } => {
+            UniformDurationMode::Large { max_secs, max_nanos, secs } => {
                 // constant folding means this is at least as fast as `gen_range`
                 let nano_range = Uniform::new(0, 1_000_000_000);
                 loop {
-                    let d = Duration::new(secs.sample(rng), nano_range.sample(rng));
-                    if d <= size {
-                        break d;
+                    let s = secs.sample(rng);
+                    let n = nano_range.sample(rng);
+                    if !(s == max_secs && n > max_nanos) {
+                        let sum = n + self.offset;
```

Comment:
After adding the offset back, `s * 1_000_000_000 + sum` is correct, but again `sum` may be more than 10^9. Again, using an `if` to check whether to decrement `sum` and increment `s` is probably the best option.